src/PowerShellEditorServices/Services/PowerShell/Host/InternalHost.cs

```diff
@@ -0,0 +1,778 @@
+                // This is more than just an optimization.
+                // When debugging, we cannot execute PowerShell directly to get this information;
+                // trying to do so will block on the command that called us, deadlocking execution.
+                // Instead, since we are reusing the runspace, we reuse that runspace's info as well.
+                (Runspace currentRunspace, RunspaceInfo currentRunspaceInfo) = _runspaceStack.Peek();
+                if (currentRunspace == pwsh.Runspace)
+                {
+                    isNewRunspace = false;
+                    return currentRunspaceInfo;
+                }
+            }
+
+            isNewRunspace = true;
+            return RunspaceInfo.CreateFromPowerShell(_logger, pwsh, _localComputerName);
+        }
+
+        private void PushPowerShellAndRunLoop(PowerShellContextFrame frame)
+        {
+            if (_psFrameStack.Count > 0)
+            {
+                RemoveRunspaceEventHandlers(CurrentFrame.PowerShell.Runspace);
+            }
+
+            AddRunspaceEventHandlers(frame.PowerShell.Runspace);
+
+            _psFrameStack.Push(frame);
+
+            try
+            {
+                if (_psFrameStack.Count == 1)
+                {
+                    RunTopLevelExecutionLoop();
+                }
+                else if ((frame.FrameType & PowerShellFrameType.Debug) != 0)
+                {
+                    RunDebugExecutionLoop();
+                }
+                else
+                {
+                    RunExecutionLoop();
+                }
+            }
+            finally
+            {
+                PopPowerShell();
+            }
+        }
+
+        private void PopPowerShell()
+        {
+            _shouldExit = false;
+            PowerShellContextFrame frame = _psFrameStack.Pop();
+            try
+            {
+                // If we're changing runspace, make sure we move the handlers over
+                if (_runspaceStack.Peek().Item1 != CurrentPowerShell.Runspace)
+                {
+                    (Runspace parentRunspace, _) = _runspaceStack.Pop();
+                    RemoveRunspaceEventHandlers(frame.PowerShell.Runspace);
+                    AddRunspaceEventHandlers(parentRunspace);
+                }
+            }
+            finally
+            {
+                frame.Dispose();
+            }
+        }
+
+        private void RunTopLevelExecutionLoop()
+        {
+            // Make sure we execute any startup tasks first
+            if (_psFrameStack.Count == 1)
+            {
+                while (_taskQueue.TryTake(out ISynchronousTask task))
+                {
+                    task.ExecuteSynchronously(CancellationToken.None);
+                }
+            }
+
+            RunExecutionLoop();
+        }
+
+        private void RunDebugExecutionLoop()
+        {
+            try
+            {
+                DebugContext.EnterDebugLoop(CancellationToken.None);
```

Comment:
One of a few cancellation tokens that we might want to decide whether to use or get rid of. My philosophy has generally been to pass them up through callers until I can't any more